### DISCUSSION TITLE
Math from scratch in Haskell: zero and one | Loominate

### DISCUSSION THREAD

**Discussion by giscus** [michaelgwelch/loominate.net#9] [Blog Comments]
# Math from scratch in Haskell: zero and one | Loominate

Follow along as I implement natural numbers from scratch in Haskell, using Eric Lippert’s posts as a guide.

http://loominate.net/2013/10/09/math-from-scratch-in-haskell-zero-and-one/

**Comment by michaelgwelch** [comment]
Hey historical @michaelgwelch, if&mdash;as you say&mdash;`Nat` adds a new bit in the least significant bit position, then your constructor for `Nat` would be a lot more intuitive if it were the following

```haskell
data Natural = Zero | Nat Natural Bit
```

instead of

```haskell
data Natural = Zero | Nat Bit Natural
```

Then when doing pattern matching and destructuring, it's obvious the the `Bit` is the least significant bit. Right now it's very confusing.

&mdash;Future @michaelgwelch
